ntp-udp/src/lib.rs

```diff
@@ -1,219 +1,424 @@
+                    }
+                } else {
+                    None
+                }
+            }
+        }
 
-#[repr(C)]
-#[allow(non_camel_case_types)]
-struct ifreq {
-    ifrn_name: [u8; 16],
-    ifru_data: *mut libc::c_void,
-    __empty_space: [u8; 40 - 8],
-}
+        /// The space used to store a control message that contains a value of type T
+        pub(crate) const fn control_message_space<T>() -> usize {
+            // Safety: CMSG_SPACE is safe to call
+            (unsafe { libc::CMSG_SPACE((std::mem::size_of::<T>()) as _) }) as usize
+        }
 
-impl TimestampingConfig {
-    /// Enable all timestamping options that are supported by this crate and the hardware/software
-    /// of the device we're running on
-    #[allow(dead_code)]
-    fn all_supported(udp_socket: &std::net::UdpSocket) -> std::io::Result<Self> {
-        // Get time stamping and PHC info
-        const ETHTOOL_GET_TS_INFO: u32 = 0x00000041;
-
-        let mut tsi: ethtool_ts_info = ethtool_ts_info {
-            cmd: ETHTOOL_GET_TS_INFO,
-            ..Default::default()
-        };
+        fn zeroed_sockaddr_storage() -> libc::sockaddr_storage {
+            // a zeroed-out sockaddr storage is semantically valid, because a ss_family with value 0 is
+            // libc::AF_UNSPEC. Hence the rest of the data does not come with any constraints
+            // Safety:
+            // the MaybeUninit is zeroed before assumed to be initialized
+            unsafe { std::mem::MaybeUninit::zeroed().assume_init() }
+        }
+    }
 
-        let fd = udp_socket.as_raw_fd();
+    mod timestamping_config {
+        use std::os::unix::prelude::AsRawFd;
 
-        if let Some(ifrn_name) = interface_name::interface_name(udp_socket.local_addr()?)? {
-            let ifr: ifreq = ifreq {
-                ifrn_name,
-                ifru_data: (&mut tsi as *mut _) as *mut libc::c_void,
-                __empty_space: [0; 40 - 8],
-            };
+        use super::cerr;
+        use crate::interface_name;
+
+        #[derive(Debug, Clone, Copy, Default)]
+        pub(crate) struct TimestampingConfig {
+            pub(crate) rx_software: bool,
+            pub(crate) tx_software: bool,
+        }
 
-            const SIOCETHTOOL: u64 = 0x8946;
-            cerr(unsafe { libc::ioctl(fd, SIOCETHTOOL, &ifr) }).unwrap();
+        #[repr(C)]
+        #[allow(non_camel_case_types)]
+        #[derive(Default)]
+        struct ethtool_ts_info {
+            cmd: u32,
+            so_timestamping: u32,
+            phc_index: u32,
+            tx_types: u32,
+            tx_reserved: [u32; 3],
+            rx_filters: u32,
+            rx_reserved: [u32; 3],
+        }
 
-            let support = Self {
-                rx_software: tsi.so_timestamping & libc::SOF_TIMESTAMPING_RX_SOFTWARE != 0,
-                tx_software: tsi.so_timestamping & libc::SOF_TIMESTAMPING_TX_SOFTWARE != 0,
-            };
+        #[repr(C)]
+        union ifr_ifru {
```

Comment:
Where is this defined?

Reply:
https://github.com/torvalds/linux/blob/master/include/uapi/linux/if.h

I'm trying to get that into the `libc` crate, because these types  are a bit nasty and we only use tiny parts of these types, yet have to implement them fully so  they are ABI-compatible.

Reply:
As long as it's not in the libc crate (though probably insofar as you're trying to upstream it), would be good to add a link to that near the definition.